# Final Project Explainer: 02806 Social Data Analysis and Visualization

This is our final project explainer notebook. We discuss our project motivation, dataset, analysis, findings, and limitations in detail.

To keep this explainer neat and readable, we keep most of the heavy analysis code in the analysis notebook:

[Analysis notebook](https://github.com/Rakibul-Islam-GitHub/Social-data-analysis-project/blob/main/notebooks/02_analysis.ipynb)

We have also published our magazine-style narrative webpage here:

[Project webpage](https://rakibul-islam-github.github.io/Social-data-analysis-project/web/)

---

## 1. Motivation
Air pollution is a major public health issue, especially in cities. Fine particulate matter, known as PM2.5, is linked to serious health problems, including cardiovascular disease, respiratory disease, lung cancer, and premature death [[1]](https://www.who.int/news-room/fact-sheets/detail/ambient-(outdoor)-air-quality-and-health). Children and people with existing health conditions are especially vulnerable because air pollution can affect lung development and worsen respiratory problems [[2]](https://www.niehs.nih.gov/health/topics/agents/air-pollution).

A common way to talk about city air pollution is to use one average value. For example, we often hear that one city is more polluted than another city. This is easy to understand, but it hides many important details. A city can have a moderate yearly average but still experience several months of high pollution. Another city can have short but intense pollution episodes. These patterns matter because exposure is not only about the average. It is also about timing, seasonality, and how often people experience unhealthy days.Recent research also supports this idea. Seasonal air-pollution patterns can change exposure levels and population vulnerability across places and time periods [[3]](https://doi.org/10.1038/s41598-025-98123-8). 

These were our primary motivations to explore these patterns. We also wanted to investigate how they relate to key weather variables such as temperature, wind speed, cloud cover, and other meteorological factors.

### 1.1 About our project idea

Initially, the project started with a narrower focus on Dubai, where we wanted to compare pollution patters across a few monitoring locations inside Dubai. However, after exploring the data and receiving feedback from our 1 min movie submitted as project assignment A, we realized that focussing purely on the Dubai story was too limited for the type of narrative we wanted to build. The different sensor locations within Dubai were not strong enough to support a full data story that will give a strong analysis to the audience. 

Instead of forcing a weak story, we expanded the project into a broader global city comparison. Dubai is still included, but it is now one city in a wider set of **15 cities** from different urban and environmental contexts. This gave us a stronger story and allowed us to compare cities through several angles, such as seasonality, daily exceedance burden, particle profile, weather context, and city clustering.

### 1.2 About our dataset

Our dataset combines two public data sources.

The first source is **OpenAQ**, which provides open access to global air-quality data. OpenAQ includes monitoring locations, sensors, pollutant measurements, and station metadata [[4]](https://docs.openaq.org/). We used OpenAQ to collect PM2.5 and PM10 data from monitoring stations across our selected cities.

The second source is **Open-Meteo**, which provides historical weather data at hourly resolution for locations around the world. We used it to add weather variables such as temperature, wind speed, humidity, precipitation, cloud cover, wind direction, and surface pressure [[5]](https://open-meteo.com/en/docs/historical-weather-api).

In the beginning, we considered 18 cities. After checking the available sensor coverage and pollutant availability, we selected **15 cities** for the final analysis. The final cities include examples from Gulf and desert urbanization, high-pollution Asian cities, dense Asian comparison cities, developed-city comparisons, and one optional contrast city.

The final analysis uses:

- Hourly OpenAQ PM2.5 and PM10 measurements
- City-level hourly air-quality data
- Daily air-quality summaries
- Monthly air-quality summaries
- Hourly Open-Meteo weather data
- Daily and monthly weather summaries

In total, we collected more than **1.5M raw hourly air-quality records** from multiple sensors across the selected cities. These raw records were later processed into city-hourly, daily, and monthly datasets.

### 1.3 Why did we choose these datasets?

First of all, we chose the air quality related dataset because air pollution is an important global issue and visually interesting at the same time. OpenAQ gives access to real monitoring-station data from many countries. This was important because our goal was to compare cities using actual sensor measurements, not only model-based estimates. OpenAQ also gave us location-level information, which allowed us to check which cities had usable PM2.5 and PM10 data.

We also chose Open-Meteo because air pollution does not happen separately from weather. Wind can move or dilute pollution. Temperature and humidity can affect how particles form or stay in the air. Cloud cover and precipitation can also help explain daily changes. Weather data helped us build a more complete story about pollution episodes.

These datasets were also suitable for a data visualization project. They gave us several time levels and comparison levels:

- Hourly patterns
- Daily pollution episodes
- Monthly seasonality
- City-to-city comparison
- Particle-profile analysis
- Weather-linked analysis

This allowed us to move beyond simple bar charts and create more advanced visuals such as maps, heatmaps, bump charts, dot matrices, ridgeline plots, episode fingerprints, and city clustering maps.

### 1.4 Main research question

Our main research question is:

**What does a city average hide about urban air pollution?**

To answer this, we break the main question into smaller questions:

1. How do PM2.5 levels differ across cities in the laste year (2025)?
2. Do city rankings stay stable across the year?
3. How often does each city exceed the WHO daily PM2.5 guideline?
4. Do cities have different PM2.5 and PM10 particle profiles?
5. How do weather conditions relate to different pollution episodes?
6. Can we group cities by their overall air-quality fingerprints instead of only ranking them by average PM2.5?

### 1.5 Goal for the end user experience

Our goal was to make the project understandable also for a non-technical audience. The website is designed as a guided visual story, while this notebook provides a more in depth explanation to the technical work behind it.

The end user should be able to scroll through the project and gradually understand the main idea:

**A city average is useful, but it is not enough. And there are many underlying things that should be known to us for our better quality of life**

By the end of the webpage, the user should understand that air pollution has several hidden layers:

- Some cities are polluted for most of the year.
- Some cities have strong seasonal changes.
- Some cities have many days above the WHO guideline.
- Some cities have more fine-particle-dominated pollution.
- Some cities have more PM10-heavy days.
- Weather can change the shape of pollution episodes.
- Cities can be compared as air-quality fingerprints, not only as rankings.

This is why our final webpage uses a magazine-style structure. We start with a global map, then move through seasonality, ranking changes, daily exceedance burden, particle profiles, weather episodes, and city fingerprints. The story is designed to help readers see patterns step by step.

### 1.6 Why this project matters

This project matters because public discussions about air pollution often focus on simple rankings. Rankings are easy to share, but they can hide important differences in how people experience pollution.

For example, two cities can have similar yearly PM2.5 levels but very different daily patterns. One city may have steady moderate pollution, while another may have shorter but more severe pollution episodes. From a health and policy point of view, these are not the same experience.

By using daily and monthly patterns, we can show a more complete picture. The project does not claim to identify exact pollution sources. That would require more detailed source data. Instead, it shows how open air-quality and weather data can be turned into a clearer visual story about exposure, timing, and city differences.



## 2. Basic Stats

In this section, we summarize the dataset that was built for the project.  
We first look at the OpenAQ data, then we check city coverage and explain why three cities were removed from the main analysis. Finally, we summarize the enriched OpenAQ plus Open-Meteo dataset used in the final visual analysis.

The raw OpenAQ data was collected at a local monitoring station level. Since each city has multiple monitoring stations, there is quite a few measurements for the same city and hour. To make the data usable for comparison, we processed it into different levels: raw hourly pollutant records, location-hourly records, city-hourly records, city-daily records, and city-monthly records.

The raw dataset contained more than 1.5M hourly records across the selected cities and sensors. We chose to use 2025 as the common analysis year because of the recency and it is the most commonly recorded year across the different city sensors. Sticking to one common year between the different cities made the comparison fairer and easier to explain.

### For city selection:

We initially considered 18 cities that had different urban and environmental contexts. After filtering through and checking data availability and quality, we selected 15 cities for the main analysis:

Bangkok, Beijing, Delhi, Dhaka, Dubai, Jakarta, Lahore, London, Los Angeles, Mexico City, New York, Riyadh, Seoul, Singapore, and Tokyo.

Abu Dhabi, Cape Town, and Kuwait City were removed from the main analysis because their coverage was not strong enough or did not align well with the 2025 comparison window.

### **Data Cleaning and Preprocessing**

Data cleaning was necessary because some of the pollutant values were clearly invalid. For example, we found extremely high values such as 9999 or values above 5000 which is not realistic. These PM2.5 measurements were likely missing-values or sensor errors. To account for a rare unfortunate problem with a sensor, we treated PM2.5 values below 0 or above 1000 as invalid, and PM10 values below 0 or above 2000 as invalid. These values were replaced with missing values before the daily and monthly datasets were rebuilt.

For many of the visual comparisons, we used the median for PM2.5 instead of mean. Median values are better to use when there are outliers from short spikes and sensor errors. When we did the WHO analysis however, we used daily mean PM2.5 because the official daily WHO guidline is based on a 24-hour mean.

In [2]:
# Basic setup

from pathlib import Path
import pandas as pd
import numpy as np
from IPython.display import display

project_root = Path(".").resolve()

# If this notebook is inside notebooks/, move one level up
if not (project_root / "data").exists():
    project_root = project_root.parent

openaq_by_city_dir = project_root / "data" / "processed" / "openaq_by_city"
enriched_by_city_dir = project_root / "data" / "processed" / "enriched" / "by_city"
enriched_combined_dir = project_root / "data" / "processed" / "enriched" / "combined"
city_selection_dir = project_root / "data" / "processed" / "combined" / "city_selection"

stats_output_dir = project_root / "data" / "processed" / "dataset_overview"
stats_output_dir.mkdir(parents=True, exist_ok=True)

In [3]:
# Small helper functions


def clean_city_key(value):
    return str(value).strip().lower().replace(" ", "_").replace("-", "_")


def find_file(city_folder, ending):
    files = sorted(city_folder.glob(f"*{ending}"))
    return files[0] if files else None


def count_csv_rows(path):
    if path is None or not Path(path).exists():
        return 0
    return sum(len(chunk) for chunk in pd.read_csv(path, chunksize=250_000))


def count_csv_rows_between(
    path,
    date_columns=("datetime_hour_utc", "local_date"),
    start="2020-01-01",
    end="2025-12-31",
):
    if path is None or not Path(path).exists():
        return 0

    path = Path(path)
    columns = pd.read_csv(path, nrows=0).columns
    date_col = next((c for c in date_columns if c in columns), None)

    if date_col is None:
        return count_csv_rows(path)

    start = pd.Timestamp(start, tz="UTC")
    end = pd.Timestamp(end, tz="UTC") + pd.Timedelta(days=1)

    total = 0
    for chunk in pd.read_csv(path, usecols=[date_col], chunksize=250_000):
        dates = pd.to_datetime(chunk[date_col], errors="coerce", utc=True)
        total += int(((dates >= start) & (dates < end)).sum())

    return total


def first_last_date(path, date_columns=("datetime_hour_utc", "local_date")):
    if path is None or not Path(path).exists():
        return pd.NaT, pd.NaT

    path = Path(path)
    columns = pd.read_csv(path, nrows=0).columns
    date_col = next((c for c in date_columns if c in columns), None)

    if date_col is None:
        return pd.NaT, pd.NaT

    first_date = pd.NaT
    last_date = pd.NaT

    for chunk in pd.read_csv(path, usecols=[date_col], chunksize=250_000):
        dates = pd.to_datetime(chunk[date_col], errors="coerce", utc=True).dropna()

        if len(dates) == 0:
            continue

        first_date = (
            dates.min() if pd.isna(first_date) else min(first_date, dates.min())
        )
        last_date = dates.max() if pd.isna(last_date) else max(last_date, dates.max())

    return first_date, last_date


def count_unique_values(path, possible_columns):
    if path is None or not Path(path).exists():
        return np.nan

    columns = pd.read_csv(path, nrows=0).columns
    use_col = next((c for c in possible_columns if c in columns), None)

    if use_col is None:
        return np.nan

    values = set()
    for chunk in pd.read_csv(path, usecols=[use_col], chunksize=250_000):
        values.update(chunk[use_col].dropna().astype(str).unique())

    return len(values)

In [4]:
# Load selected and excluded city lists


def load_city_keys(path):
    path = Path(path)

    if not path.exists():
        return set()

    df = pd.read_csv(path)
    city_col = "city_key" if "city_key" in df.columns else "city"

    if city_col not in df.columns:
        return set()

    return set(df[city_col].map(clean_city_key))


selected_keys = load_city_keys(city_selection_dir / "selected_cities_pm25_main.csv")
excluded_keys = load_city_keys(city_selection_dir / "excluded_or_context_cities.csv")


def city_status(city_key):
    if city_key in selected_keys:
        return "selected main city"
    if city_key in excluded_keys:
        return "excluded or context city"
    return "candidate city"


print("Selected main cities:", len(selected_keys))
print("Excluded or context cities:", len(excluded_keys))

Selected main cities: 15
Excluded or context cities: 3


In [5]:
# 2.1 OpenAQ raw and processed file summary for all candidate cities

city_folders = sorted([p for p in openaq_by_city_dir.iterdir() if p.is_dir()])

rows = []

for city_folder in city_folders:
    city_key = clean_city_key(city_folder.name)

    raw_file = find_file(city_folder, "_openaq_hourly_long.csv")
    location_hourly_file = find_file(city_folder, "_location_hourly.csv")
    city_hourly_file = find_file(city_folder, "_city_hourly.csv")
    city_daily_file = find_file(city_folder, "_city_daily.csv")
    city_monthly_file = find_file(city_folder, "_city_monthly.csv")

    first_hour, last_hour = first_last_date(city_hourly_file, ("datetime_hour_utc",))
    first_day, last_day = first_last_date(city_daily_file, ("local_date",))

    rows.append(
        {
            "city_key": city_key,
            "status": city_status(city_key),

            "raw_hourly_rows_2020_2025": count_csv_rows_between(
                raw_file, ("datetime_hour_utc", "datetime"), "2020-01-01", "2025-12-31"
            ),
            "location_hourly_rows": count_csv_rows(location_hourly_file),
            "city_hourly_rows": count_csv_rows(city_hourly_file),
            "city_daily_rows": count_csv_rows(city_daily_file),
            "city_monthly_rows": count_csv_rows(city_monthly_file),
            "first_hour": first_hour,
            "last_hour": last_hour,
            "first_day": first_day,
            "last_day": last_day,
        }
    )

openaq_city_summary = pd.DataFrame(rows)

openaq_city_summary.to_csv(
    stats_output_dir / "openaq_city_file_summary.csv", index=False
)

display(openaq_city_summary.sort_values("raw_hourly_rows_2020_2025", ascending=False))

,city_key,status,raw_hourly_rows_2020_2025,location_hourly_rows,city_hourly_rows,city_daily_rows,city_monthly_rows,first_hour,last_hour,first_day,last_day
11,los_angeles,selected main city,219107,20587,19243,901,34,2022-02-18 01:00:00+00:00,2026-01-01 07:00:00+00:00,2022-02-17 00:00:00+00:00,2025-12-31 00:00:00+00:00
10,london,selected main city,213560,120003,33520,1464,55,2021-05-11 00:00:00+00:00,2026-01-01 00:00:00+00:00,2021-05-11 00:00:00+00:00,2026-01-01 00:00:00+00:00
4,delhi,selected main city,208077,40121,18659,1058,44,2019-12-31 19:00:00+00:00,2025-12-31 18:00:00+00:00,2019-12-31 00:00:00+00:00,2025-12-31 00:00:00+00:00
1,bangkok,selected main city,176306,92870,33587,1529,55,2021-04-29 19:00:00+00:00,2025-12-31 17:00:00+00:00,2021-04-29 00:00:00+00:00,2025-12-31 00:00:00+00:00
15,seoul,selected main city,95161,47708,15715,666,24,2024-02-28 01:00:00+00:00,2025-12-31 15:00:00+00:00,2024-02-28 00:00:00+00:00,2026-01-01 00:00:00+00:00
13,new_york,selected main city,81099,53793,16819,739,27,2023-11-25 06:00:00+00:00,2026-01-01 05:00:00+00:00,2023-11-25 00:00:00+00:00,2026-01-01 00:00:00+00:00
17,tokyo,selected main city,78284,77622,20156,887,31,2023-07-27 16:00:00+00:00,2025-12-31 15:00:00+00:00,2023-07-28 00:00:00+00:00,2026-01-01 00:00:00+00:00
12,mexico_city,selected main city,69586,37296,16720,766,33,2020-10-17 14:00:00+00:00,2026-01-01 06:00:00+00:00,2020-10-17 00:00:00+00:00,2026-01-01 00:00:00+00:00
2,beijing,selected main city,62157,35042,23293,1058,39,2019-12-31 17:00:00+00:00,2025-12-31 16:00:00+00:00,2020-01-01 00:00:00+00:00,2026-01-01 00:00:00+00:00
9,lahore,selected main city,60385,16279,15172,763,26,2023-11-28 16:00:00+00:00,2025-12-31 19:00:00+00:00,2023-11-28 00:00:00+00:00,2025-12-31 00:00:00+00:00


In [6]:
print("Raw OpenAQ hourly records, 2020 to 2025:")
print(
    f"All candidate cities: {openaq_city_summary['raw_hourly_rows_2020_2025'].sum():,}"
)


Raw OpenAQ hourly records, 2020 to 2025:
All candidate cities: 1,534,411


In [4]:
# Load selected and excluded city lists


def load_city_keys(path):
    path = Path(path)

    if not path.exists():
        return set()

    df = pd.read_csv(path)
    city_col = "city_key" if "city_key" in df.columns else "city"

    if city_col not in df.columns:
        return set()

    return set(df[city_col].map(clean_city_key))


selected_keys = load_city_keys(city_selection_dir / "selected_cities_pm25_main.csv")
excluded_keys = load_city_keys(city_selection_dir / "excluded_or_context_cities.csv")


def city_status(city_key):
    if city_key in selected_keys:
        return "selected main city"
    if city_key in excluded_keys:
        return "excluded or context city"
    return "candidate city"


print("Selected main cities:", len(selected_keys))
print("Excluded or context cities:", len(excluded_keys))

Selected main cities: 15
Excluded or context cities: 3


In [5]:
# 2.1 OpenAQ raw and processed file summary for all candidate cities

city_folders = sorted([p for p in openaq_by_city_dir.iterdir() if p.is_dir()])

rows = []

for city_folder in city_folders:
    city_key = clean_city_key(city_folder.name)

    raw_file = find_file(city_folder, "_openaq_hourly_long.csv")
    location_hourly_file = find_file(city_folder, "_location_hourly.csv")
    city_hourly_file = find_file(city_folder, "_city_hourly.csv")
    city_daily_file = find_file(city_folder, "_city_daily.csv")
    city_monthly_file = find_file(city_folder, "_city_monthly.csv")

    first_hour, last_hour = first_last_date(city_hourly_file, ("datetime_hour_utc",))
    first_day, last_day = first_last_date(city_daily_file, ("local_date",))

    rows.append(
        {
            "city_key": city_key,
            "status": city_status(city_key),

            "raw_hourly_rows_2020_2025": count_csv_rows_between(
                raw_file, ("datetime_hour_utc", "datetime"), "2020-01-01", "2025-12-31"
            ),
            "location_hourly_rows": count_csv_rows(location_hourly_file),
            "city_hourly_rows": count_csv_rows(city_hourly_file),
            "city_daily_rows": count_csv_rows(city_daily_file),
            "city_monthly_rows": count_csv_rows(city_monthly_file),
            "first_hour": first_hour,
            "last_hour": last_hour,
            "first_day": first_day,
            "last_day": last_day,
        }
    )

openaq_city_summary = pd.DataFrame(rows)

openaq_city_summary.to_csv(
    stats_output_dir / "openaq_city_file_summary.csv", index=False
)

display(openaq_city_summary.sort_values("raw_hourly_rows_2020_2025", ascending=False))

,city_key,status,raw_hourly_rows_2020_2025,location_hourly_rows,city_hourly_rows,city_daily_rows,city_monthly_rows,first_hour,last_hour,first_day,last_day
11,los_angeles,selected main city,219107,20587,19243,901,34,2022-02-18 01:00:00+00:00,2026-01-01 07:00:00+00:00,2022-02-17 00:00:00+00:00,2025-12-31 00:00:00+00:00
10,london,selected main city,213560,120003,33520,1464,55,2021-05-11 00:00:00+00:00,2026-01-01 00:00:00+00:00,2021-05-11 00:00:00+00:00,2026-01-01 00:00:00+00:00
4,delhi,selected main city,208077,40121,18659,1058,44,2019-12-31 19:00:00+00:00,2025-12-31 18:00:00+00:00,2019-12-31 00:00:00+00:00,2025-12-31 00:00:00+00:00
1,bangkok,selected main city,176306,92870,33587,1529,55,2021-04-29 19:00:00+00:00,2025-12-31 17:00:00+00:00,2021-04-29 00:00:00+00:00,2025-12-31 00:00:00+00:00
15,seoul,selected main city,95161,47708,15715,666,24,2024-02-28 01:00:00+00:00,2025-12-31 15:00:00+00:00,2024-02-28 00:00:00+00:00,2026-01-01 00:00:00+00:00
13,new_york,selected main city,81099,53793,16819,739,27,2023-11-25 06:00:00+00:00,2026-01-01 05:00:00+00:00,2023-11-25 00:00:00+00:00,2026-01-01 00:00:00+00:00
17,tokyo,selected main city,78284,77622,20156,887,31,2023-07-27 16:00:00+00:00,2025-12-31 15:00:00+00:00,2023-07-28 00:00:00+00:00,2026-01-01 00:00:00+00:00
12,mexico_city,selected main city,69586,37296,16720,766,33,2020-10-17 14:00:00+00:00,2026-01-01 06:00:00+00:00,2020-10-17 00:00:00+00:00,2026-01-01 00:00:00+00:00
2,beijing,selected main city,62157,35042,23293,1058,39,2019-12-31 17:00:00+00:00,2025-12-31 16:00:00+00:00,2020-01-01 00:00:00+00:00,2026-01-01 00:00:00+00:00
9,lahore,selected main city,60385,16279,15172,763,26,2023-11-28 16:00:00+00:00,2025-12-31 19:00:00+00:00,2023-11-28 00:00:00+00:00,2025-12-31 00:00:00+00:00


In [6]:
print("Raw OpenAQ hourly records, 2020 to 2025:")
print(
    f"All candidate cities: {openaq_city_summary['raw_hourly_rows_2020_2025'].sum():,}"
)


Raw OpenAQ hourly records, 2020 to 2025:
All candidate cities: 1,534,411


From OpenAQ, we collected 1,534,411 raw hourly pollutant measurement records from 2020 to 2025 across the 18 candidate cities. These raw records were in long format, where PM2.5 and PM10 measurements could appear as separate rows for the same monitoring location and hour.

After cleaning and reshaping, the raw records were converted into location-hourly data. In this processed version, each row represents one monitoring location at one hour, with PM2.5 and PM10 stored as separate columns when available. This produced more than 700,000 location-hour observations. These were then aggregated into city-hourly, city-daily, and city-monthly datasets for analysis.

In [7]:
print("Raw OpenAQ hourly after cleaning & reshaping, 2020 to 2025:")
print(f"All candidate cities: {openaq_city_summary['location_hourly_rows'].sum():,}")

Raw OpenAQ hourly after cleaning & reshaping, 2020 to 2025:
All candidate cities: 743,462


In [8]:
# 2.3 Load all city-daily OpenAQ files to check year coverage

daily_frames = []

for city_folder in city_folders:
    city_key = clean_city_key(city_folder.name)
    city_daily_file = find_file(city_folder, "_city_daily.csv")

    if city_daily_file is None:
        continue

    cols = pd.read_csv(city_daily_file, nrows=0).columns.tolist()

    needed_cols = [
        c
        for c in [
            "city",
            "country",
            "city_group",
            "local_date",
            "pm25_daily_mean",
            "pm25_daily_median",
            "pm10_daily_mean",
            "pm10_daily_median",
            "valid_pm25_hours",
            "valid_pm10_hours",
        ]
        if c in cols
    ]

    df = pd.read_csv(city_daily_file, usecols=needed_cols)
    df["city_key"] = city_key
    daily_frames.append(df)

all_city_daily = pd.concat(daily_frames, ignore_index=True)

all_city_daily["local_date"] = pd.to_datetime(
    all_city_daily["local_date"], errors="coerce"
)
all_city_daily["year"] = all_city_daily["local_date"].dt.year
all_city_daily["month"] = all_city_daily["local_date"].dt.month
all_city_daily["status"] = all_city_daily["city_key"].map(city_status)

for col in [
    "pm25_daily_mean",
    "pm10_daily_mean",
    "valid_pm25_hours",
    "valid_pm10_hours",
]:
    if col in all_city_daily.columns:
        all_city_daily[col] = pd.to_numeric(all_city_daily[col], errors="coerce")

all_city_daily["pm25_valid_day"] = all_city_daily["pm25_daily_mean"].notna() & (
    all_city_daily["valid_pm25_hours"].fillna(0) >= 12
)

all_city_daily["pm10_valid_day"] = all_city_daily["pm10_daily_mean"].notna() & (
    all_city_daily["valid_pm10_hours"].fillna(0) >= 12
)

all_city_daily = all_city_daily[all_city_daily["year"].between(2020, 2025)].copy()

print("All candidate city-daily rows, 2020 to 2025:", len(all_city_daily))
display(all_city_daily.head())

All candidate city-daily rows, 2020 to 2025: 15641


,city,country,city_group,local_date,pm25_daily_mean,pm25_daily_median,valid_pm25_hours,pm10_daily_mean,pm10_daily_median,valid_pm10_hours,city_key,year,month,status,pm25_valid_day,pm10_valid_day
0,abu_dhabi,NaN,NaN,2020-01-02,34.525000,31.35,16,NaN,NaN,0,abu_dhabi,2020,1,excluded or context city,True,False
1,abu_dhabi,NaN,NaN,2020-01-03,27.656250,25.00,16,NaN,NaN,0,abu_dhabi,2020,1,excluded or context city,True,False
2,abu_dhabi,NaN,NaN,2020-01-05,17.066667,13.55,24,NaN,NaN,0,abu_dhabi,2020,1,excluded or context city,True,False
3,abu_dhabi,NaN,NaN,2020-01-07,6.787500,6.55,24,NaN,NaN,0,abu_dhabi,2020,1,excluded or context city,True,False
4,abu_dhabi,NaN,NaN,2020-01-14,13.913043,11.50,23,NaN,NaN,0,abu_dhabi,2020,1,excluded or context city,True,False


Year coverage table

In [9]:
# 2.4 Year coverage table

year_city_coverage = (
    all_city_daily.groupby(["year", "city_key", "status"], dropna=False)
    .agg(
        days_with_any_data=("local_date", "nunique"),
        valid_pm25_days=("pm25_valid_day", "sum"),
        valid_pm10_days=("pm10_valid_day", "sum"),
        months_with_any_data=("month", "nunique"),
    )
    .reset_index()
)

year_city_coverage["has_6_months_pm25"] = year_city_coverage["valid_pm25_days"] >= 180
year_city_coverage["has_9_months_pm25"] = year_city_coverage["valid_pm25_days"] >= 270

year_summary = (
    year_city_coverage.groupby("year")
    .agg(
        candidate_cities_with_any_data=("city_key", "nunique"),
        cities_with_180_plus_pm25_days=("has_6_months_pm25", "sum"),
        cities_with_270_plus_pm25_days=("has_9_months_pm25", "sum"),
        total_valid_pm25_days=("valid_pm25_days", "sum"),
        total_valid_pm10_days=("valid_pm10_days", "sum"),
        average_pm25_days_per_city=("valid_pm25_days", "mean"),
    )
    .reset_index()
)

year_summary.to_csv(
    stats_output_dir / "year_coverage_summary_2020_2025.csv", index=False
)

display(year_summary)

,year,candidate_cities_with_any_data,cities_with_180_plus_pm25_days,cities_with_270_plus_pm25_days,total_valid_pm25_days,total_valid_pm10_days,average_pm25_days_per_city
0,2020,5,3,2,1190,669,238.000000
1,2021,8,3,1,1277,968,159.625000
2,2022,8,3,0,1159,709,144.875000
3,2023,12,4,4,1948,1010,162.333333
4,2024,16,10,10,4040,2446,252.500000
5,2025,16,15,14,5044,1713,315.250000


In [10]:
# 2.5 City-level 2025 coverage table to justify final city selection
# This version keeps all 18 candidate cities, including cities with no usable 2025 rows.

# Candidate city list comes from the OpenAQ city folders
candidate_cities = pd.DataFrame(
    {"city_key": [clean_city_key(p.name) for p in city_folders]}
).drop_duplicates()

candidate_cities["status"] = candidate_cities["city_key"].map(city_status)

# Base 2025 coverage from available daily rows
coverage_2025_available = year_city_coverage[year_city_coverage["year"] == 2025].copy()

coverage_2025_available = coverage_2025_available.drop(
    columns=["status"], errors="ignore"
)

# Count valid PM2.5 months in 2025
pm25_months_2025 = (
    all_city_daily[
        (all_city_daily["year"] == 2025) & (all_city_daily["pm25_valid_day"])
    ]
    .groupby("city_key")["month"]
    .nunique()
    .rename("valid_pm25_months")
)

# Count valid PM10 months in 2025
pm10_months_2025 = (
    all_city_daily[
        (all_city_daily["year"] == 2025) & (all_city_daily["pm10_valid_day"])
    ]
    .groupby("city_key")["month"]
    .nunique()
    .rename("valid_pm10_months")
)

# Start from all candidate cities so missing cities remain visible
coverage_2025 = (
    candidate_cities.merge(coverage_2025_available, on="city_key", how="left")
    .merge(pm25_months_2025, on="city_key", how="left")
    .merge(pm10_months_2025, on="city_key", how="left")
)

# Fill missing coverage values with zero
coverage_columns = [
    "year",
    "days_with_any_data",
    "valid_pm25_days",
    "valid_pm10_days",
    "months_with_any_data",
    "valid_pm25_months",
    "valid_pm10_months",
]

for col in coverage_columns:
    if col not in coverage_2025.columns:
        coverage_2025[col] = 0

coverage_2025["year"] = coverage_2025["year"].fillna(2025).astype(int)
coverage_2025["days_with_any_data"] = (
    coverage_2025["days_with_any_data"].fillna(0).astype(int)
)
coverage_2025["valid_pm25_days"] = (
    coverage_2025["valid_pm25_days"].fillna(0).astype(int)
)
coverage_2025["valid_pm10_days"] = (
    coverage_2025["valid_pm10_days"].fillna(0).astype(int)
)
coverage_2025["months_with_any_data"] = (
    coverage_2025["months_with_any_data"].fillna(0).astype(int)
)
coverage_2025["valid_pm25_months"] = (
    coverage_2025["valid_pm25_months"].fillna(0).astype(int)
)
coverage_2025["valid_pm10_months"] = (
    coverage_2025["valid_pm10_months"].fillna(0).astype(int)
)

# Add selection decision
coverage_2025["decision_note"] = np.where(
    coverage_2025["status"] == "selected main city",
    "kept in main analysis",
    "removed from main comparison or kept as context",
)

# Add a simple reason column
coverage_2025["coverage_reason"] = "usable 2025 PM2.5 coverage"

coverage_2025.loc[coverage_2025["valid_pm25_days"] == 0, "coverage_reason"] = (
    "no valid PM2.5 daily data in 2025"
)

coverage_2025.loc[
    (coverage_2025["valid_pm25_days"] > 0) & (coverage_2025["valid_pm25_months"] < 10),
    "coverage_reason",
] = "not enough valid PM2.5 months in 2025"

coverage_2025.loc[
    coverage_2025["status"] == "selected main city", "coverage_reason"
] = "kept because 2025 PM2.5 coverage was usable"

# Sort: excluded/context first, then selected cities
coverage_2025 = coverage_2025.sort_values(
    ["status", "valid_pm25_months", "valid_pm25_days"], ascending=[True, False, False]
).reset_index(drop=True)

coverage_2025.to_csv(
    stats_output_dir / "city_coverage_2025_selection_check.csv", index=False
)

print("Total candidate cities shown:", coverage_2025["city_key"].nunique())
print("Selected main cities:", (coverage_2025["status"] == "selected main city").sum())
print(
    "Excluded/context cities:", (coverage_2025["status"] != "selected main city").sum()
)

display(coverage_2025.head(5))

Total candidate cities shown: 18
Selected main cities: 15
Excluded/context cities: 3


,city_key,status,year,days_with_any_data,valid_pm25_days,valid_pm10_days,months_with_any_data,has_6_months_pm25,has_9_months_pm25,valid_pm25_months,valid_pm10_months,decision_note,coverage_reason
0,kuwait_city,excluded or context city,2025,63,63,0,3,False,False,3,0,removed from main comparison or kept as context,not enough valid PM2.5 months in 2025
1,abu_dhabi,excluded or context city,2025,0,0,0,0,NaN,NaN,0,0,removed from main comparison or kept as context,no valid PM2.5 daily data in 2025
2,cape_town,excluded or context city,2025,0,0,0,0,NaN,NaN,0,0,removed from main comparison or kept as context,no valid PM2.5 daily data in 2025
3,london,selected main city,2025,365,365,365,12,True,True,12,12,kept in main analysis,kept because 2025 PM2.5 coverage was usable
4,seoul,selected main city,2025,365,365,365,12,True,True,12,12,kept in main analysis,kept because 2025 PM2.5 coverage was usable


In [11]:
# 2.6  cities that were excluded or kept as context

excluded_coverage_2025 = coverage_2025[
    coverage_2025["status"] != "selected main city"
].copy()

display(excluded_coverage_2025)

,city_key,status,year,days_with_any_data,valid_pm25_days,valid_pm10_days,months_with_any_data,has_6_months_pm25,has_9_months_pm25,valid_pm25_months,valid_pm10_months,decision_note,coverage_reason
0,kuwait_city,excluded or context city,2025,63,63,0,3,False,False,3,0,removed from main comparison or kept as context,not enough valid PM2.5 months in 2025
1,abu_dhabi,excluded or context city,2025,0,0,0,0,NaN,NaN,0,0,removed from main comparison or kept as context,no valid PM2.5 daily data in 2025
2,cape_town,excluded or context city,2025,0,0,0,0,NaN,NaN,0,0,removed from main comparison or kept as context,no valid PM2.5 daily data in 2025


The table shows that the exclusion was mainly a coverage decision, not a city-importance decision. Cities with no or very weak 2025 coverage were removed, while cities with partial but usable coverage were kept and interpreted with coverage-aware metrics.

In [12]:
# 2.7 Enriched OpenAQ plus Open-Meteo data overview

enriched_hourly_files = sorted(
    enriched_by_city_dir.glob("*/*_city_hourly_enriched.csv")
)

enriched_rows = []

for path in enriched_hourly_files:
    city_key = clean_city_key(path.parent.name)
    first_hour, last_hour = first_last_date(path, ("datetime_hour_utc",))

    enriched_rows.append(
        {
            "city_key": city_key,
            "status": city_status(city_key),
            "enriched_city_hourly_rows": count_csv_rows(path),
            "first_hour": first_hour,
            "last_hour": last_hour,
        }
    )

enriched_hourly_summary = pd.DataFrame(enriched_rows)

daily_enriched_path = enriched_combined_dir / "combined_city_daily_enriched_2025.csv"
monthly_enriched_path = (
    enriched_combined_dir / "combined_city_monthly_enriched_2025.csv"
)

daily_enriched = pd.read_csv(daily_enriched_path)
monthly_enriched = pd.read_csv(monthly_enriched_path)

enriched_overview = pd.DataFrame(
    [
        {
            "dataset": "Enriched city-hourly",
            "files": len(enriched_hourly_files),
            "rows": int(enriched_hourly_summary["enriched_city_hourly_rows"].sum()),
            "description": "City-hourly OpenAQ data joined with Open-Meteo weather features.",
        },
        {
            "dataset": "Enriched city-daily 2025",
            "files": 1,
            "rows": len(daily_enriched),
            "description": "One row per city per day, used for exceedance, particle profile, and episode analysis.",
        },
        {
            "dataset": "Enriched city-monthly 2025",
            "files": 1,
            "rows": len(monthly_enriched),
            "description": "One row per city per month, used for map, heatmap, and ranking analysis.",
        },
    ]
)

enriched_hourly_summary.to_csv(
    stats_output_dir / "enriched_city_hourly_summary.csv", index=False
)

enriched_overview.to_csv(
    stats_output_dir / "enriched_dataset_overview.csv", index=False
)

display(enriched_overview)
display(
    enriched_hourly_summary.sort_values("enriched_city_hourly_rows", ascending=False)
)

,dataset,files,rows,description
0,Enriched city-hourly,15,115587,City-hourly OpenAQ data joined with Open-Meteo...
1,Enriched city-daily 2025,1,5103,"One row per city per day, used for exceedance,..."
2,Enriched city-monthly 2025,1,178,"One row per city per month, used for map, heat..."


,city_key,status,enriched_city_hourly_rows,first_hour,last_hour
7,london,selected main city,8755,2025-01-01 00:00:00+00:00,2025-12-31 23:00:00+00:00
12,seoul,selected main city,8721,2025-01-01 00:00:00+00:00,2025-12-31 15:00:00+00:00
14,tokyo,selected main city,8674,2025-01-01 00:00:00+00:00,2025-12-31 15:00:00+00:00
5,jakarta,selected main city,8099,2025-01-01 00:00:00+00:00,2025-12-31 14:00:00+00:00
0,bangkok,selected main city,8094,2025-01-01 00:00:00+00:00,2025-12-31 17:00:00+00:00
9,mexico_city,selected main city,7979,2025-01-01 00:00:00+00:00,2025-12-31 23:00:00+00:00
13,singapore,selected main city,7924,2025-01-01 00:00:00+00:00,2025-12-31 16:00:00+00:00
10,new_york,selected main city,7923,2025-01-01 00:00:00+00:00,2025-12-31 23:00:00+00:00
4,dubai,selected main city,7920,2025-01-01 00:00:00+00:00,2025-12-31 20:00:00+00:00
1,beijing,selected main city,7629,2025-02-11 18:00:00+00:00,2025-12-31 16:00:00+00:00


In [13]:
# 2.8 Weather feature completeness in the enriched daily dataset

weather_columns = [
    "temperature_2m_mean_c",
    "relative_humidity_2m_mean_pct",
    "precipitation_total_mm",
    "wind_speed_10m_mean_ms",
    "wind_direction_10m_circular_mean_deg",
    "surface_pressure_mean_hpa",
    "cloud_cover_mean_pct",
]

weather_columns = [c for c in weather_columns if c in daily_enriched.columns]

weather_completeness = pd.DataFrame(
    {
        "weather_feature": weather_columns,
        "non_missing_rows": [
            int(daily_enriched[c].notna().sum()) for c in weather_columns
        ],
        "missing_rows": [int(daily_enriched[c].isna().sum()) for c in weather_columns],
        "coverage_percent": [
            round(daily_enriched[c].notna().mean() * 100, 1) for c in weather_columns
        ],
    }
)

weather_completeness.to_csv(
    stats_output_dir / "weather_feature_completeness_daily_2025.csv", index=False
)

display(weather_completeness)

,weather_feature,non_missing_rows,missing_rows,coverage_percent
0,temperature_2m_mean_c,5103,0,100.0
1,relative_humidity_2m_mean_pct,5103,0,100.0
2,precipitation_total_mm,5103,0,100.0
3,wind_speed_10m_mean_ms,5103,0,100.0
4,wind_direction_10m_circular_mean_deg,5103,0,100.0
5,surface_pressure_mean_hpa,5103,0,100.0
6,cloud_cover_mean_pct,5103,0,100.0


In [17]:
# 2.9 Preview of the final enriched daily dataset

daily_enriched_path = enriched_combined_dir / "combined_city_daily_enriched_2025.csv"

daily_enriched = pd.read_csv(daily_enriched_path)

preview_columns = [
    "city",
    "local_date",
    "pm25_daily_mean",
    "pm25_daily_median",
    "pm10_daily_mean",
    "pm10_daily_median",
    "valid_pm25_hours",
    "valid_pm10_hours",
    "temperature_2m_mean_c",
    "relative_humidity_2m_mean_pct",
    "wind_speed_10m_mean_ms",
    "cloud_cover_mean_pct",
]

preview_columns = [col for col in preview_columns if col in daily_enriched.columns]

display(daily_enriched[preview_columns].head())


,city,local_date,pm25_daily_mean,pm25_daily_median,pm10_daily_mean,pm10_daily_median,valid_pm25_hours,valid_pm10_hours,temperature_2m_mean_c,relative_humidity_2m_mean_pct,wind_speed_10m_mean_ms,cloud_cover_mean_pct
0,bangkok,2025-01-01,36.502917,36.345,55.625000,55.5,24,16,27.487500,51.875000,1.082500,27.833333
1,bangkok,2025-01-02,38.828437,38.300,63.750000,62.0,24,16,28.012500,47.458333,1.291667,37.208333
2,bangkok,2025-01-03,40.270833,40.250,73.937500,76.0,24,16,28.150000,48.875000,1.728750,49.500000
3,bangkok,2025-01-04,39.234687,38.850,67.687500,65.5,24,16,27.887500,45.750000,1.815000,58.375000
4,bangkok,2025-01-05,45.387604,45.600,76.266667,76.0,24,15,26.945833,43.708333,1.498750,84.958333


In [ ]:
# 2.10 compact table for overall dataset overview
import jinja2

basic_stats_summary = pd.DataFrame(
    [
        {
            "item": "Candidate cities considered",
            "value": len(city_folders),
            "note": "Initial city folders collected from OpenAQ processing.",
        },
        {
            "item": "Final selected cities",
            "value": len(selected_keys),
            "note": "Cities kept for the main 2025 comparison.",
        },
        {
            "item": "Excluded or context cities",
            "value": len(excluded_keys),
            "note": "Cities removed from the main comparison because of coverage or alignment issues.",
        },
        {
            "item": "Raw OpenAQ hourly rows, 2020 to 2025",
            "value": int(openaq_city_summary["raw_hourly_rows_2020_2025"].sum()),
            "note": "Raw PM2.5 and PM10 measurement records from selected monitoring sensors.",
        },
        {
            "item": "OpenAQ city-daily rows",
            "value": int(openaq_city_summary["city_daily_rows"].sum()),
            "note": "Daily city-level rows before weather enrichment.",
        },
        {
            "item": "Enriched city-hourly rows",
            "value": int(enriched_hourly_summary["enriched_city_hourly_rows"].sum()),
            "note": "City-hourly air quality joined with Open-Meteo weather data.",
        },
        {
            "item": "Final enriched daily rows, 2025",
            "value": len(daily_enriched),
            "note": "Main dataset for daily analysis.",
        },
        {
            "item": "Final enriched monthly rows, 2025",
            "value": len(monthly_enriched),
            "note": "Main dataset for monthly maps, heatmaps, and rankings.",
        },
    ]
)

basic_stats_summary.to_csv(
    stats_output_dir / "basic_stats_summary_for_explainer.csv", index=False
)
display(
    basic_stats_summary.style.set_properties(
        subset=["note"], **{"text-align": "left", "white-space": "normal"}
    ).set_table_styles(
        [
            {"selector": "th", "props": [("text-align", "left")]},
            {"selector": "td", "props": [("vertical-align", "top")]},
        ]
    )
)


,item,value,note
0,Candidate cities considered,18,Initial city folders collected from OpenAQ processing.
1,Final selected cities,15,Cities kept for the main 2025 comparison.
2,Excluded or context cities,3,Cities removed from the main comparison because of coverage or alignment issues.
3,"Raw OpenAQ hourly rows, 2020 to 2025",1534411,Raw PM2.5 and PM10 measurement records from selected monitoring sensors.
4,OpenAQ city-daily rows,15650,Daily city-level rows before weather enrichment.
5,Enriched city-hourly rows,115587,City-hourly air quality joined with Open-Meteo weather data.
6,"Final enriched daily rows, 2025",5103,Main dataset for daily analysis.
7,"Final enriched monthly rows, 2025",178,"Main dataset for monthly maps, heatmaps, and rankings."


This summary shows the scale of the data pipeline. We started with 18 candidate cities and collected 1,534,411 raw OpenAQ hourly pollutant records from 2020 to 2025. These raw rows represent PM2.5 and PM10 measurement records from monitoring sensors, so one location and one hour can appear more than once if both pollutants were measured.

After cleaning and aggregation, the raw data was converted into city-level datasets. The city-daily dataset before weather enrichment contains 15,650 rows, which gave us a broader view of coverage across years. **However, because cities did not all have the same time coverage, we selected 2025 as the common analysis window.**

The final analysis dataset is smaller but more consistent. It contains 5,103 enriched daily rows and 178 enriched monthly rows for the selected cities in 2025. These final datasets include both air-quality variables and Open-Meteo weather features, and they were used for the main visual analysis in the project.

## 3. Data Analysis

The main analytical idea of the project is that city averages hide important variation. A city with a moderate to low yearly average can still have many high-pollution days, while another city may have a high average but a more predictable seasonal pattern. To find out what the air pollution situation in these cities really was, we analyzed the data from several perspectives instead of relying on a single summary statistic.

### Monthly PM2.5 comparison
First, we compared monthly median PM2.5 across cities. This allowed us to see broad patterns throughout the year and compare cities across the same 2025 time window. Monthly medians were used because they summarize typical pollution levels while reducing the influence of extreme short-term spikes.

### City-month heatmap

The city-month heatmap was used to reveal seasonality. Instead of showing only which cities are more polluted on average, the heatmap shows when pollution is higher or lower during the year. This helps identify whether a city’s air-quality problem is stable, seasonal, or episode-driven.

### Monthly ranking transitions

 We then ranked cities by monthly median PM2.5. The purpose of this analysis was to show that rankings are not fixed. Some cities remain consistently high, while others move up or down depending on the month. This supports the main argument that a single annual ranking can hide important dynamics that change throughout the year.

### WHO exceedance analysis

To connect the analysis to health relevance, we calculated how often each city exceeded the WHO daily PM2.5 guideline. For this part, we used daily mean PM2.5, since the guideline is defined using a 24-hour mean. The exceedance dot matrix makes the health burden visible and shows just how many days this guidline is breached.

### Particle-profile analysis

PM2.5 and PM10 describe different particle-size ranges. To investigate this, we calculated the PM2.5 / PM10 ratio for city-days where both pollutants were available. A higher ratio suggests that fine particles dominate, while a lower ratio suggests a larger contribution from coarse particles. This helped us compare not only how polluted cities are, but also what type of particle profile they tend to have.

### Weather-episode fingerprint

We added on to the air-quality data by including weather variables from Open-Meteo. Weather was not used to prove causation, but to add context to why pollution episodes/extremeties take place. For example, wind speed and particle ratios can help distinguish between different types of unusually extreme pollution. The weather-episode fingerprint visualizes a combination of wind speed, PM2.5 / PM10 ratio, PM10 level, and episode type in one plot.

### City fingerprint map

Finally, we created a city fingerprint map using multiple features, including pollution burden, exceedance frequency, seasonality, particle profile, and weather-linked factors. The goal was to group cities based on their overall behaviour in air-quality rather than taking the simple one-dimensional approach of ranking them by average PM2.5.

## 4. Genre

Our website follows a magazine-style interactive data story. It is not a random dashboard where the reader has to find the story alone. It is also not a fully linear slideshow where the reader only moves from one fixed slide to the next. Instead, the page guides the reader through a clear argument, while still allowing interaction inside the visualizations.

In terms of Segel and Heer’s narrative visualization genres, our project follows **magazine-style narrative** design style [[6]](https://doi.org/10.1109/tvcg.2010.179). The page is organized into sections with headings, short explanations, captions, and embedded interactive charts. The reader scrolls through the story in a planned order, but can still hover, zoom, pan, and inspect details inside the Plotly visualizations.

The structure also follows a loose **martini glass** pattern. At the beginning, we guide the reader strongly through the main message: city averages hide important pollution patterns. As the story continues, the reader gets more room to explore individual cities, months, rankings, particle profiles, weather episodes, and city fingerprints. This creates a balance between an author-driven story and reader-driven exploration.

### 4.1 Visual Narrative Tools

Segel and Heer divide visual narrative tools into three main groups: visual structuring, highlighting, and transition guidance. We used all three in different ways.

#### Visual structuring

We used visual structuring to help the reader understand where they are in the story. The website has a clear vertical menu on the left side with section links such as Global map, Seasonality, Ranking shifts, Exceedance burden, Particle profiles, Weather episodes, and City fingerprints. This works like a table of contents and helps the reader move between sections.

Each section also has a similar structure: a title, a short explanation, a visualization, and a caption or interpretation. This repeated layout creates a consistent visual platform. The reader does not need to relearn the page structure for every new chart.

#### Highlighting

We used highlighting to direct attention to important patterns. For example, color is used to show PM2.5 intensity in the map and heatmap. In the WHO exceedance matrix, red shows days above the WHO daily PM2.5 guideline, blue shows below-guideline days, and gray shows missing or low-coverage days. In the particle-profile and weather-episode charts, marker color and shape help separate different types of pollution days.

We also used marker size in several visualizations. In the animated map, larger markers help show city-level presence. In the city fingerprint map, marker size shows WHO exceedance burden. This helps the reader quickly see which cities have a heavier pollution burden.

#### Transition guidance

The website uses section ordering and repeated layout to guide transitions. The story begins with a broad global view, then moves to monthly seasonality, then rankings, then daily exceedance burden, then particle profiles, weather episodes, and finally city fingerprints. This order helps the reader move from simple comparisons to more detailed interpretations.

We also use smooth scrolling, section headers, captions, and a fixed menu to reduce confusion between sections. The reader can always see the main story path and return to earlier parts if needed.

### 4.2 Narrative Structure

Segel and Heer divide narrative structure into ordering, interactivity, and messaging. We used each of these in the final webpage.

#### Ordering

The story has a mostly linear order. We start with the main question: **What does a city average hide?** Then we answer it step by step.

The order is:

1. Global map
2. City-month seasonality heatmap
3. Monthly ranking bump chart
4. WHO daily exceedance matrix
5. Particle-profile ridgeline
6. Weather-episode fingerprints
7. City fingerprint map
8. Final takeaways

This ordering was chosen because it slowly builds the argument. First, the reader sees where pollution is high. Then they see that pollution changes by month. Then they see that rankings change. Then they see that daily exceedance burden matters. Finally, they see that cities have different pollution fingerprints.

#### Interactivity

The webpage uses limited interactivity. The reader can hover over charts to see exact values, explore different months in the animated map, inspect city-level points, and compare details across cities. This keeps the page interactive without turning it into a complex dashboard.

We chose limited interactivity because our target audience is non-technical. Too many filters or controls could distract from the main story. The goal was to let the reader explore details while still following a clear narrative.

#### Messaging

Messaging is one of the most important parts of our story. Each section has explanatory text before the visualization and a short interpretation after it. We also use clear section titles, captions, and final takeaways.

The main message is repeated throughout the project: **A city average is useful, but it is not enough.**


## 5. Visualizations

This section explains the main visualizations used in our project. The figures shown here are static PNG versions so that the explainer notebook can be viewed easily on GitHub. The full code used to create these visualizations is available in the [Analysis notebook](https://github.com/Rakibul-Islam-GitHub/Social-data-analysis-project/blob/main/notebooks/02_analysis.ipynb)

Most of the original charts are interactive figures. For the full user experience, including hover details, animation, zooming, and map interaction, please visit our [magazine-style webpag](https://rakibul-islam-github.github.io/Social-data-analysis-project/web/)

### 5.1 Global Monthly PM2.5 Animated Map

![Monthly median PM2.5 map](../outputs/figures_png/autoplay_monthly_median_pm25_global_map.png)

This map gives the first global overview of our selected cities. Each circle represents one city in the final comparison. The color of the circle shows the monthly median PM2.5 concentration, while the marker size reflects the number of active monitoring locations used for that city and month.

The map is animated by month, from January 2025 to December 2025. This helps us see that air pollution is not fixed across the year. Some cities stay relatively high for many months, while others change more strongly by season. The map also gives the reader a quick sense of the geographic spread of the project.

In this figure, in January view, several Asian and Gulf cities already show stronger PM2.5 levels than many developed-city comparisons. This sets up the main question of the project: if pollution differs so much across space and time, what does a single city average hide?

#### Why we chose this visualization

- It gives the reader an immediate global overview of the selected cities.
- It shows both location and pollution intensity in one figure.
- The monthly animation helps show that pollution changes across the year.
- Marker size adds a simple coverage cue, so the reader can see that cities are not all based on the same number of active monitoring sensor locations.

### 5.2 City-month heatmap

![City-month median PM2.5 heatmap](../outputs/figures_png/city_month_median_pm25_heatmap.png)

This heatmap shows monthly PM2.5 for each selected city in 2025. Each row is a city and each column is a month. Darker colors mean higher PM2.5 levels.

The chart shows that pollution is not constant across the year. Delhi, Dhaka, and Lahore have several months with very high PM2.5, while cities such as New York, London, Los Angeles, and Tokyo stay much lower. Some cities also show clear seasonal changes. For example, Bangkok is higher early in the year and again in December, while Beijing and Dubai show more moderate but changing patterns.

This figure supports one of our main points: a yearly city average hides seasonal differences. Two cities can have similar yearly values but very different month-by-month patterns.

#### Why we chose this visualization

- It shows city and month patterns in one compact view.
- It makes seasonality easy to compare across cities.
- It helps identify cities with repeated high-pollution months.
- It works better than a line chart because it avoids many overlapping city lines.
- It supports our argument that annual averages are not enough.

### 5.3 Montly ranking bump chart

![Monthly PM2.5 ranking bump chart](../outputs/figures_png/monthly_pm25_median_bump_chart.png)

This bump chart ranks the selected cities by monthly median PM2.5. Rank 1 means the highest PM2.5 level for that month.

The chart shows that city rankings are not fixed. Delhi, Dhaka, and Lahore often stay near the top, but their positions still change across the year. Other cities move more strongly. Bangkok starts high early in the year, falls in the middle months, and rises again in December. Dubai also moves between middle and higher ranks depending on the month.

This chart supports our argument that a single yearly ranking is too simple. The answer to “which city is most polluted?” depends on the month.

#### Why we chose this visualization

- It shows ranking changes more clearly than a normal line chart.
- It makes monthly movement easy to follow.
- It highlights that pollution rankings are dynamic.
- It works well after the heatmap because it turns monthly values into relative city positions.

### 5.4 WHO exceedance dot matrix

![WHO daily PM2.5 exceedance matrix](../outputs/figures_png/fig_exceedance_dot_matrix.png)

This chart shows daily PM2.5 exceedance. Each small square represents one city-day. Red means the daily mean PM2.5 was above the WHO 24-hour guideline of 15 µg/m³. Blue means the day was below the level and it's safe, and gray means the day had low coverage or missing data.

The chart shows a clear difference between cities. Delhi, Lahore, and Dhaka exceed the WHO guideline on almost all valid days. Jakarta and Dubai also show very frequent exceedance. In contrast, New York, Tokyo, London, and Los Angeles have many more below-guideline days, although they still have some exceedance days.

This chart is important because it moves the story from averages to lived exposure. A city with many red days means people are exposed to unhealthy PM2.5 levels repeatedly, even if the yearly average alone does not show the full pattern.

#### Why we chose this visualization

- It shows daily exposure more clearly than an annual average.
- It makes missing or low-coverage days visible instead of hiding them.
- It directly connects the analysis to the WHO daily PM2.5 guideline.
- The side bar chart gives a quick summary of exceedance burden.
- It helps readers compare both timing and frequency of polluted days.

### 5.5 Particle-profile ridgeline

![Particle profile ridgeline plot](../outputs/figures_png/particle_profile_ridgeline_plot.png)

This ridgeline plot shows the daily PM2.5 / PM10 ratio for cities with enough paired PM2.5 and PM10 data. Each ridge shows the distribution of daily ratios for one city. The black tick marks each city’s median ratio.

The PM2.5 / PM10 ratio helps describe the particle profile. A higher ratio means PM2.5 makes up a larger share of PM10, so the day is more fine-particle dominated. A lower ratio means PM10 is relatively larger, which can suggest more coarse-particle influence. This is only a descriptive indicator, not direct source attribution.

The chart shows that cities do not have the same particle profile. New York has the highest ratio in this filtered paired dataset, while Delhi, Bangkok, London, Seoul, and Mexico City are closer to the 0.5 reference line. The highlighted markers show high-pollution days with weather context, helping us see that bad-air days can have different particle patterns.

#### Why we chose this visualization

- It shows the full distribution instead of only one average ratio.
- It helps compare particle profiles across cities.
- It makes fine-particle and PM10-heavy patterns easier to see.
- It adds weather-linked episode markers without making a separate chart.
- It supports the idea that pollution is not only about how much PM2.5 a city has, but also about the particle mix.

### 5.6 Weather-episode fingerprint

![Weather-episode fingerprint plot](../outputs/figures_png/weather_episode_fingerprint.png)

This small-multiple chart connects pollution episodes with weather context. Each point is one city-day. The x-axis shows wind speed, the y-axis shows the PM2.5 / PM10 ratio, marker size represents PM10, and color shows the episode type.

The chart shows that high-pollution days do not happen in the same way everywhere. Delhi has many fine-particle high-PM2.5 days at lower to moderate wind speeds, while PM10-heavy low-ratio days appear lower on the y-axis. Bangkok and Mexico City show more compact clusters. London has a wider wind-speed range, with many lower-pollution days. New York has fewer paired valid days, so it should be interpreted carefully.

The dotted lines show each city’s median PM2.5 / PM10 ratio and 75th-percentile wind speed. Points to the right of the vertical line are relatively windy days for that city. Points below the horizontal line are more PM10-heavy compared with that city’s usual particle profile.

#### Why we chose this visualization

- It combines air quality, particle profile, and weather in one view.
- It shows that high-pollution days can have different conditions behind them.
- The small-multiple layout makes city-to-city comparison easier.
- It helps separate fine-particle high-PM2.5 days from PM10-heavy days.
- It supports the idea that air pollution should be read as episodes, not only as averages.

### 5.7 City fingerprint map

![City fingerprint PCA map](../outputs/figures_png/city_fingerprint_pca.png)

This chart summarizes each city using several features together, not just one PM2.5 value. The features include PM2.5 burden, exceedance frequency, seasonality, particle profile, and weather context. The PCA projection places cities with similar overall patterns closer together.

Delhi, Dhaka, and Lahore form a clear high-burden cluster on the right side. They have large marker sizes, which means they have a high share of days above the WHO daily PM2.5 guideline. Dubai and Riyadh appear separately in the upper part of the chart, suggesting a different Gulf city pattern. Lower-burden cities such as New York, London, Tokyo, Los Angeles, Seoul, Singapore, and Mexico City sit more toward the left side.

The arrows show which features shape the fingerprint space. PM2.5, seasonality, and volatility point toward the high-burden cluster, while weather variables such as wind, humidity, and cloud cover help separate cities along the second axis.

#### Why we chose this visualization

- It combines several city features into one summary view.
- It shows that cities can be grouped by overall behavior, not only ranked by PM2.5.
- It helps reveal clear clusters, such as South Asian high-burden cities.
- Marker size adds the WHO exceedance burden to the same figure.

## 6. Discussion

This section reflects on what worked well in the project, what was the limitations, and what could be improved if we had more data.

### 6.1 What went well

Overall, the project became stronger after we changed direction from a Dubai-only study to a global city comparison. At first, we planned to focus on a few monitoring locations inside Dubai. After checking the data, we saw that the Dubai location-level differences were not strong enough for a full story. Instead of forcing that idea, we expanded the project to 15 cities around the world. This gave us a richer dataset and a better story.

The data pipeline also worked well in the end. We collected more than 1.5 million raw OpenAQ pollutant records from 2020 to 2025, reshaped them into location-hourly data, then aggregated them into city-hourly, daily, and monthly datasets. This helped us move from messy sensor-level data to clean analysis-ready data.

The coverage audit was another important success. We did not simply choose cities and assume they all had good data. We checked which cities had enough PM2.5 data, which had PM10 data, and which cities had paired PM2.5 and PM10. This helped us remove Abu Dhabi, Cape Town, and Kuwait City from the main comparison because their 2025 coverage was too weak. This made the final analysis more fair.

The use of weather data also improved the project. By adding Open-Meteo variables such as temperature, wind speed, humidity, cloud cover, and precipitation, we could look at pollution in a more detailed way. Weather did not prove the cause of pollution, but it helped us describe different pollution episodes more clearly.

The final visual story also worked well. Instead of relying on simple bar charts, we used several visual forms:

- Animated map for global overview
- Heatmap for seasonality
- Bump chart for ranking changes
- Dot matrix for daily WHO exceedance burden
- Ridgeline plot for PM2.5 / PM10 particle profiles
- Small-multiple scatter plots for weather episodes
- PCA fingerprint map for overall city comparison

Together, these charts support the main message of the project: a city average hides many important pollution patterns.

The webpage also helped make the project more readable for a non-technical audience. The vertical menu, section layout, captions, and interactive charts made the story easier to follow than a normal notebook or report.

### 6.2 What we learned from the data

The project showed that air pollution cannot be explained well with one average value. Some cities are consistently high, while others change strongly across the year. The heatmap and bump chart showed that rankings are not fixed. A city can move up or down depending on the month.

The WHO exceedance matrix was especially useful because it changed the focus from average pollution to daily exposure. Cities such as Delhi, Lahore, and Dhaka exceeded the WHO daily PM2.5 guideline on almost all valid days. Other cities had lower yearly levels but still experienced some above-guideline days.

The PM2.5 / PM10 ratio also showed that pollution is not the same everywhere. Some cities had more fine-particle-dominated profiles, while others had relatively more PM10-heavy days. This helped us treat pollution as a pattern, not only as a number.

The PCA fingerprint map gave a final summary of the analysis. It showed that Delhi, Dhaka, and Lahore form a clear high-burden group, while Dubai and Riyadh behave differently from both the South Asian cities and the lower-burden developed-city group. This supports the idea that cities have air-quality fingerprints.

### 6.3 What was difficult

The hardest part of the project was data collection and cleaning. OpenAQ data is very useful, but it is not evenly available for every city. Some cities have many sensors, some have only a few, and not every sensor measures both PM2.5 and PM10. The time coverage also differs between cities.

Another challenge was processing the S3 files. Fetching all data for every monitoring location was slow, especially for cities with many sensors. We had to make practical decisions, such as selecting a smaller number of useful monitoring locations per city.

Invalid values were also a problem. For example, we found unrealistic PM values such as 9999 in Seoul. These values would have distorted the results if we kept them. Cleaning them was necessary before doing the final analysis.

Another challenge was balancing detail and readability. The dataset had hourly, daily, monthly, pollutant, weather, and city-level information. It was easy to make charts that were technically rich but hard to read. We had to choose visualizations that were advanced but still understandable.

### 6.4 What is still missing

The biggest limitation is that we did not do source attribution. The PM2.5 / PM10 ratio can suggest whether a day is more fine-particle dominated or more PM10-heavy, but it cannot prove the exact pollution source. For example, we cannot say for sure whether a PM10-heavy day was caused by dust, traffic, construction, or industry. To do that, we would need more data, such as emissions inventories, land-use data, road networks, industrial locations.

Another limitation is sensor coverage. Even after filtering, not all selected cities have perfect coverage. Some cities have full-year coverage, while others have partial coverage. Dhaka, for example, has fewer valid PM2.5 days than cities such as London or Seoul. We handled this by using coverage-aware calculations, but the difference in coverage still matters.

The project also uses city-level aggregation. This is useful for comparison, but it hides within-city variation. A large city can have very different pollution levels across neighborhoods. Since our final story compares cities, we do not fully explore neighborhood-level inequality or exposure differences within cities.

Weather data was added using representative city-level coordinates. This is practical, but it is not perfect. Weather can vary inside large cities, and pollution sensors may not always be located exactly where the weather point is. A more detailed project could use sensor-level weather matching.

The PCA fingerprint map is useful as an exploratory view, but it should not be interpreted as a strict model. It helps group cities by similar features, but the result depends on which features are included, how they are scaled, and how missing values are handled.

### 6.5 What could be improved

If we had more time, we would improve the project in several ways.

First, we would expand the data coverage. We could test more cities and more monitoring locations, then build a stronger selection rule. This would make the city comparison more balanced.

Second, we would improve sensor selection. In this project, we selected a manageable number of locations per city to keep the S3 fetching practical. A future version could create a more formal location-selection method using distance from city center, coverage percentage, pollutant availability, and spatial spread.

Third, we would add more source-related data. Land-use data, traffic density, desert dust indicators, industrial zones data could help explain why cities have different particle profiles.

Fourth, we would improve the webpage further. The current webpage works as a magazine-style visual story, but it could include more scroll-based animations, stronger annotations, and clearer transitions between sections. Some charts could also be simplified for mobile screens.

Finally, we would add uncertainty and coverage indicators more directly into the charts. We already use valid-day counts and missing-day colors in some figures, but this could be made even clearer throughout the story.

### 6.6 Final reflection

The main strength of this project is that it does not stop at a simple pollution ranking. It shows that air pollution has many layers: seasonality, daily exceedance, particle profile, weather context, and city-level fingerprints.

The project also shows how much work is needed before visualization can begin. Most of the effort went into collecting, checking, cleaning, reshaping, and combining data. Without that work, the final charts would not be trustworthy.

Our final story is not perfect, but it gives a more complete view than a single average PM2.5 value. It shows that cities should be compared not only by how polluted they are, but also by when pollution happens, how often people experience unhealthy days, and what kind of particle profile the city has.

## 7. Contributions

 **Rakibul Islam (s260126):** Responsible for project scaling idea from 3 cities to 15 cities by worldwide data collection, cleaning and merging of final datasets, selection of the cities globally, finalized data analysis & processing, generated a big portion of the plots, created the website, helped in writing explainer, web content.

 **Victor Eriksen (s234667):** Responsible for initial idea and initial data collection, cleaning & merging of initial data, finding the first OpenAQ sensors & Open-Meteo locations, generating the first few plots, writing text for the plots and data story, writing a big portion of the explainer notebook, helped in different technical decision.

## 8. References

[1] World Health Organization, “Ambient (outdoor) air pollution,” Oct. 2024. [Online]. Available: https://www.who.int/news-room/fact-sheets/detail/ambient-(outdoor)-air-quality-and-health

[2] National Institute of Environmental Health Sciences, “Air Pollution and Your Health.” [Online]. Available: https://www.niehs.nih.gov/health/topics/agents/air-pollution

[3] B. Vandeninden *et al.*, “Implications of spatial and seasonal air pollution patterns, socioeconomic disparities, and 15-minute communities for achieving WHO air quality guidelines,” *Sci. Rep.*, vol. 15, p. 13683, 2025, doi: 10.1038/s41598-025-98123-8.

[4] OpenAQ, “OpenAQ API documentation.” [Online]. Available: https://docs.openaq.org/

[5] Open-Meteo, “Free Open-Source Weather API.” [Online]. Available: https://open-meteo.com/

[6] E. Segel and J. Heer, “Narrative visualization: Telling stories with data,” *IEEE Trans. Vis. Comput. Graph.*, vol. 16, no. 6, pp. 1139–1148, Nov./Dec. 2010, doi: 10.1109/TVCG.2010.179.